## 4. Preprocesamiento con Librerías
Con el uso de librerías, realice en Python los mismos preprocesamientos del punto 3.


## Ajustes generales

In [32]:
from google.colab import drive
import pandas as pd

# Montar Google Drive
drive.mount('/content/drive')

# Cargar el archivo CSV
df = pd.read_csv('/content/drive/My Drive/TAREAS U/8vo semestre/IA/Apoyo 1er parcial/Spotify_Dataset_V3.csv', delimiter=';')


df['Artists'] = df['Artists'].str.split('-')


print(df.head())



Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
   Rank            Title                       Artists        Date  \
0     1  Ella Baila Sola  [Eslabon Armado, Peso Pluma]  29/05/2023   
1     1  Ella Baila Sola  [Eslabon Armado, Peso Pluma]  29/05/2023   
2     2   WHERE SHE GOES                   [Bad Bunny]  29/05/2023   
3     3  La Bebe - Remix       [Yng Lvcas, Peso Pluma]  29/05/2023   
4     3  La Bebe - Remix       [Yng Lvcas, Peso Pluma]  29/05/2023   

   Danceability  Energy  Loudness  Speechiness  Acousticness  \
0         0.668   0.758   -5176.0        0.033         0.483   
1         0.668   0.758   -5176.0        0.033         0.483   
2         0.652   0.800   -4019.0        0.061         0.143   
3         0.812   0.479   -5678.0        0.333         0.213   
4         0.812   0.479   -5678.0        0.333         0.213   

   Instrumentalness  Valence # of Artist   Artist (Ind.) # of Nat

In [33]:
def detalles(df):
    # Número de columnas
    numero_columnas = df.shape[1]

    # Nombres de las columnas
    nombres_columnas = df.columns.tolist()

    # Número de filas
    numero_filas = df.shape[0]

    print(f"Número de columnas: {numero_columnas}")
    print("Nombres de las columnas:", nombres_columnas)
    print(f"Número de filas: {numero_filas}")


detalles(df)

Número de columnas: 20
Nombres de las columnas: ['Rank', 'Title', 'Artists', 'Date', 'Danceability', 'Energy', 'Loudness', 'Speechiness', 'Acousticness', 'Instrumentalness', 'Valence', '# of Artist', 'Artist (Ind.)', '# of Nationality', 'Nationality', 'Continent', 'Points (Total)', 'Points (Ind for each Artist/Nat)', 'id', 'Song URL']
Número de filas: 23181


In [34]:
#Mejoras

# Convertir la columna 'Date' a tipo datetime
df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y')

# Crear un diccionario para mapear los nombres de los artistas a números
artist_mapping = {
    'Artist 1': 1,
    'Artist 2': 2,
    'Artist 3': 3,
    'Artist 4': 4,
    'Artist 5': 5
}

# Usar el método replace para transformar la columna
df['Num_of_Artists'] = df['# of Artist'].replace(artist_mapping)

# Verificar el contenido de la nueva columna
print(df[['# of Artist', 'Num_of_Artists']].head())

# Dropear la columna '# of Artist'
df = df.drop(columns=['# of Artist'])
# Crear un diccionario para mapear las nacionalidades a números
nationality_mapping = {
    'Nationality 1': 1,
    'Nationality 2': 2,
    'Nationality 3': 3,
    'Nationality 4': 4,
    'Nationality 5': 5
}

# Usar el método replace para transformar la columna
df['Num_of_Nationality'] = df['# of Nationality'].replace(nationality_mapping)

# Verificar el contenido de la nueva columna
print(df[['# of Nationality', 'Num_of_Nationality']].head())

# Dropear la columna '# of Nationality'
df = df.drop(columns=['# of Nationality'])

#DECIDI ELIMNAR TITULO Y MI ARRYA DE NOMBRE ARTISTAS
df = df.drop(columns=['Title'])
df = df.drop(columns=['Artists'])


# Eliminar las columnas 'id' y 'Song URL'
df = df.drop(columns=['id', 'Song URL'])



  # of Artist  Num_of_Artists
0    Artist 1               1
1    Artist 2               2
2    Artist 1               1
3    Artist 1               1
4    Artist 2               2
  # of Nationality  Num_of_Nationality
0    Nationality 1                   1
1    Nationality 2                   2
2    Nationality 1                   1
3    Nationality 1                   1
4    Nationality 2                   2


<ipython-input-34-5b0eb35f01c1>:16: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Num_of_Artists'] = df['# of Artist'].replace(artist_mapping)
<ipython-input-34-5b0eb35f01c1>:33: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Num_of_Nationality'] = df['# of Nationality'].replace(nationality_mapping)


In [35]:
detalles(df)
print(df.head())

Número de columnas: 16
Nombres de las columnas: ['Rank', 'Date', 'Danceability', 'Energy', 'Loudness', 'Speechiness', 'Acousticness', 'Instrumentalness', 'Valence', 'Artist (Ind.)', 'Nationality', 'Continent', 'Points (Total)', 'Points (Ind for each Artist/Nat)', 'Num_of_Artists', 'Num_of_Nationality']
Número de filas: 23181
   Rank       Date  Danceability  Energy  Loudness  Speechiness  Acousticness  \
0     1 2023-05-29         0.668   0.758   -5176.0        0.033         0.483   
1     1 2023-05-29         0.668   0.758   -5176.0        0.033         0.483   
2     2 2023-05-29         0.652   0.800   -4019.0        0.061         0.143   
3     3 2023-05-29         0.812   0.479   -5678.0        0.333         0.213   
4     3 2023-05-29         0.812   0.479   -5678.0        0.333         0.213   

   Instrumentalness  Valence   Artist (Ind.)  Nationality      Continent  \
0             0.000    0.834  Eslabon Armado       Mexico  Latin-America   
1             0.000    0.834      

In [36]:
# Función para mostrar información sobre columnas categóricas
def explorar_columnas_categoricas(df):
    for col in df.select_dtypes(include=['object']).columns:
        if isinstance(df[col].iloc[0], list):  # Ignora columnas que contienen listas
            continue

        num_totales = len(df[col])  # Número total de valores
        num_unicos = df[col].nunique()  # Número de valores únicos
        valores_unicos = df[col].unique()  # Valores únicos

        print(f"Columna: {col}")
        print(f"Número total de valores: {num_totales}")
        print(f"Número de valores únicos: {num_unicos}")
        print(f"Valores únicos (primeros 10): {valores_unicos[:10]}")  # Muestra solo los primeros 10 valores únicos
        print("---")

# Explorar columnas categóricas
explorar_columnas_categoricas(df)

Columna: Artist (Ind.)
Número total de valores: 23181
Número de valores únicos: 336
Valores únicos (primeros 10): ['Eslabon Armado' 'Peso Pluma' 'Bad Bunny' 'Yng Lvcas' 'FIFTY FIFTY'
 'Grupo Frontera' 'Miley Cyrus' 'David Kushner' 'SZA' 'Harry Styles']
---
Columna: Nationality
Número total de valores: 23181
Número de valores únicos: 36
Valores únicos (primeros 10): ['Mexico' 'Puerto Rico' 'South Korea' 'United States' 'United Kingdom'
 'Colombia' 'Nigeria' 'Canada' 'Spain' 'France']
---
Columna: Continent
Número total de valores: 23181
Número de valores únicos: 7
Valores únicos (primeros 10): ['Latin-America' 'Asia' 'Anglo-America' 'Europe' 'Africa' 'Oceania'
 'Unknown']
---


In [37]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split


## OneHotEncoder
Trata cada categoría como independiente, sin asumir que hay un orden o jerarquía entre ellas.

In [38]:
ohe = OneHotEncoder(sparse_output=False)

# Aplicar OneHotEncoder
encoded_features = ohe.fit_transform(df[categorical_cols])

# Crear un DataFrame con las columnas codificadas
encoded_df = pd.DataFrame(encoded_features, columns=ohe.get_feature_names_out(categorical_cols))

df = pd.concat([df.drop(columns=categorical_cols), encoded_df], axis=1)

print(df.head())


   Rank       Date  Danceability  Energy  Loudness  Speechiness  Acousticness  \
0     1 2023-05-29         0.668   0.758   -5176.0        0.033         0.483   
1     1 2023-05-29         0.668   0.758   -5176.0        0.033         0.483   
2     2 2023-05-29         0.652   0.800   -4019.0        0.061         0.143   
3     3 2023-05-29         0.812   0.479   -5678.0        0.333         0.213   
4     3 2023-05-29         0.812   0.479   -5678.0        0.333         0.213   

   Instrumentalness  Valence  Points (Total)  ...  Nationality_United Kingdom  \
0             0.000    0.834             200  ...                         0.0   
1             0.000    0.834             200  ...                         0.0   
2             0.629    0.234             199  ...                         0.0   
3             0.000    0.559             198  ...                         0.0   
4             0.000    0.559             198  ...                         0.0   

   Nationality_United Stat

## Label Encoding
 Label Encoding asigna un número entero a cada categoría, lo que puede crear una relación ordinal falsa entre ellas. Por ejemplo, "Artist 1" podría ser 0 y "Artist 2" 1, sugiriendo que "Artist 2" es "mejor" o "mayor" que "Artist 1", lo cual no es cierto.

## Discretización de Variables

Proceso de convertir variables numéricas continuas en categorías discretas o "bins".Nos ayuda a interpretar y tambien a identificar patrones.
 En este caso, vamos a discretizar las variables **Danceability** y **Energy** por las siguientes razones:

   - **Danceability**: Indica cuán adecuada es una canción para bailar.
   - **Energy**: Esta variable refleja la intensidad y actividad de una canción.


In [39]:
# Discretización para 'Danceability' y 'Energy'
# bins 5 , osea  5 categorias
df['Danceability_binned'] = pd.cut(df['Danceability'], bins=5, labels=False)
df['Energy_binned'] = pd.cut(df['Energy'], bins=5, labels=False)

In [40]:
print(df[['Danceability', 'Danceability_binned', 'Energy', 'Energy_binned']].head())


   Danceability  Danceability_binned  Energy  Energy_binned
0         0.668                    3   0.758              3
1         0.668                    3   0.758              3
2         0.652                    3   0.800              3
3         0.812                    4   0.479              2
4         0.812                    4   0.479              2


In [41]:
print("Distribución de 'Danceability_binned':")
print(df['Danceability_binned'].value_counts())

print("Distribución de 'Energy_binned':")
print(df['Energy_binned'].value_counts())


Distribución de 'Danceability_binned':
Danceability_binned
3    9471
2    8051
4    3779
1    1729
0     151
Name: count, dtype: int64
Distribución de 'Energy_binned':
Energy_binned
3    10506
2     7130
4     3595
1     1621
0      329
Name: count, dtype: int64


### Resultados:

#### **1. Danceability:**
- **Bin 3**: Con **9471** valores, es el más común, indicando que muchas canciones tienen bailabilidad media-alta.
- **Bin 2**: **8051** valores sugieren una buena cantidad de canciones con bailabilidad baja-media.
- **Bines extremos (0 y 1)**: Representan pocos valores, indicando que hay pocas canciones con baja bailabilidad.

#### **2. Energy:**
- **Bin 3**: Con **10506** valores, la mayoría de las canciones tienen energía media-alta.
- **Bines extremos (0 y 1)**: Similar a Danceability, hay pocos ejemplos de canciones con baja energía.

#### **Interpretación General:**
- **Tendencia General**: Ambas distribuciones muestran que la mayoría de las canciones tienen una "bailabilidad" y "energía" media a alta, lo que puede reflejar una tendencia en la música popular actual.
- **Menor Representación de Bins Extremas**: Las canciones con características extremas (muy bajas o muy altas) son menos frecuentes.



## Normalización
Técnica escalar las características numéricas de un dataset a un rango similar. Al normalizar los datos, se reduce la influencia de las diferentes escalas de las variables, lo que puede mejorar la convergencia y el rendimiento del modelo.

In [42]:
numeric_cols = ['Danceability', 'Energy', 'Loudness', 'Speechiness', 'Acousticness',
                'Instrumentalness', 'Valence', 'Points (Total)',
                'Points (Ind for each Artist/Nat)', 'Num_of_Artists', 'Num_of_Nationality']

scaler = StandardScaler()
df[numeric_cols] = scaler.fit_transform(df[numeric_cols])

# Mostrar los resultados de la normalización
print(df[numeric_cols].head())

   Danceability    Energy  Loudness  Speechiness  Acousticness  \
0     -0.060460  0.681445 -0.029892    -0.670052      1.047175   
1     -0.060460  0.681445 -0.029892    -0.670052      1.047175   
2     -0.181634  0.963016  0.377901    -0.364715     -0.414239   
3      1.030105 -1.188990 -0.206826     2.601422     -0.113359   
4      1.030105 -1.188990 -0.206826     2.601422     -0.113359   

   Instrumentalness   Valence  Points (Total)  \
0         -0.171931  1.333454        1.654353   
1         -0.171931  1.333454        1.654353   
2          7.023328 -1.240858        1.637321   
3         -0.171931  0.153561        1.620290   
4         -0.171931  0.153561        1.620290   

   Points (Ind for each Artist/Nat)  Num_of_Artists  Num_of_Nationality  
0                          0.637714       -0.599147           -0.599147  
1                          0.637714        1.002979            1.002979  
2                          2.706076       -0.599147           -0.599147  
3           

In [43]:
print(df.head())

   Rank       Date  Danceability    Energy  Loudness  Speechiness  \
0     1 2023-05-29     -0.060460  0.681445 -0.029892    -0.670052   
1     1 2023-05-29     -0.060460  0.681445 -0.029892    -0.670052   
2     2 2023-05-29     -0.181634  0.963016  0.377901    -0.364715   
3     3 2023-05-29      1.030105 -1.188990 -0.206826     2.601422   
4     3 2023-05-29      1.030105 -1.188990 -0.206826     2.601422   

   Acousticness  Instrumentalness   Valence  Points (Total)  ...  \
0      1.047175         -0.171931  1.333454        1.654353  ...   
1      1.047175         -0.171931  1.333454        1.654353  ...   
2     -0.414239          7.023328 -1.240858        1.637321  ...   
3     -0.113359         -0.171931  0.153561        1.620290  ...   
4     -0.113359         -0.171931  0.153561        1.620290  ...   

   Nationality_Unknown  Continent_Africa  Continent_Anglo-America  \
0                  0.0               0.0                      0.0   
1                  0.0               0